

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Reading VCF &#8212; sgkit  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/typehints.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/docsearch.sbt.css" />
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'vcf';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.15.2';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = '_static/switcher.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'latest';
        DOCUMENTATION_OPTIONS.show_version_warning_banner = false;
        </script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
    <script defer="defer" src="_static/docsearch.sbt.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Examples" href="examples/index.html" />
    <link rel="prev" title="User Guide" href="user_guide.html" />

  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>



  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/sgkit_trnsprnt.png" class="logo__image only-light" alt="sgkit  documentation - Home"/>
    <script>document.write(`<img src="_static/sgkit_blue_trnsprnt.png" class="logo__image only-dark" alt="sgkit  documentation - Home"/>`);</script>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="user_guide.html">
                        User Guide
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        Reading VCF
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="examples/index.html">
                        Examples
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="api.html">
                        API reference
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="how_do_i.html">
                        How do I …
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="contributing.html">
                        Contributing to sgkit
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="about.html">
                        About
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="news.html">
                        News
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="changelog.html">
                        Changelog
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">
<script>
document.write(`
  <div class="version-switcher__container dropdown">
    <button id="pst-version-switcher-button-2"
      type="button"
      class="version-switcher__button btn btn-sm navbar-btn dropdown-toggle"
      data-bs-toggle="dropdown"
      aria-haspopup="listbox"
      aria-controls="pst-version-switcher-list-2"
      aria-label="Version switcher list"
    >
      Choose version  <!-- this text may get changed later by javascript -->
      <span class="caret"></span>
    </button>
    <div id="pst-version-switcher-list-2"
      class="version-switcher__menu dropdown-menu list-group-flush py-0"
      role="listbox" aria-labelledby="pst-version-switcher-button-2">
      <!-- dropdown will be populated by javascript on page load -->
    </div>
  </div>
`);
</script></div>
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/sgkit-dev/sgkit" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="getting_started.html">
                        Getting Started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="user_guide.html">
                        User Guide
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        Reading VCF
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="examples/index.html">
                        Examples
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="api.html">
                        API reference
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="how_do_i.html">
                        How do I …
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="contributing.html">
                        Contributing to sgkit
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="about.html">
                        About
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="news.html">
                        News
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="changelog.html">
                        Changelog
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<script>
document.write(`
  <div class="version-switcher__container dropdown">
    <button id="pst-version-switcher-button-3"
      type="button"
      class="version-switcher__button btn btn-sm navbar-btn dropdown-toggle"
      data-bs-toggle="dropdown"
      aria-haspopup="listbox"
      aria-controls="pst-version-switcher-list-3"
      aria-label="Version switcher list"
    >
      Choose version  <!-- this text may get changed later by javascript -->
      <span class="caret"></span>
    </button>
    <div id="pst-version-switcher-list-3"
      class="version-switcher__menu dropdown-menu list-group-flush py-0"
      role="listbox" aria-labelledby="pst-version-switcher-button-3">
      <!-- dropdown will be populated by javascript on page load -->
    </div>
  </div>
`);
</script></div>
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/sgkit-dev/sgkit" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"></div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page">Reading VCF</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
<section id="reading-vcf">
<span id="vcf"></span><h1>Reading VCF<a class="headerlink" href="#reading-vcf" title="Permalink to this heading">#</a></h1>
<nav class="contents local" id="table-of-contents">
<p class="topic-title">Table of contents:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#highlights" id="id1">Highlights</a></p></li>
<li><p><a class="reference internal" href="#installation" id="id2">Installation</a></p></li>
<li><p><a class="reference internal" href="#usage" id="id3">Usage</a></p></li>
<li><p><a class="reference internal" href="#multiple-files" id="id4">Multiple files</a></p></li>
<li><p><a class="reference internal" href="#partitioning" id="id5">Partitioning</a></p></li>
<li><p><a class="reference internal" href="#chunk-sizes" id="id6">Chunk sizes</a></p></li>
<li><p><a class="reference internal" href="#cloud-storage" id="id7">Cloud storage</a></p></li>
<li><p><a class="reference internal" href="#compression" id="id8">Compression</a></p></li>
<li><p><a class="reference internal" href="#low-level-operation" id="id9">Low-level operation</a></p></li>
<li><p><a class="reference internal" href="#polyploid-and-mixed-ploidy-vcf" id="id10">Polyploid and mixed-ploidy VCF</a></p></li>
<li><p><a class="reference internal" href="#example-converting-1000-genomes-vcf-to-zarr" id="id11">Example: converting 1000 genomes VCF to Zarr</a></p>
<ul>
<li><p><a class="reference internal" href="#install-sgkit" id="id12">Install sgkit</a></p></li>
<li><p><a class="reference internal" href="#download-the-data" id="id13">Download the data</a></p></li>
<li><p><a class="reference internal" href="#run-the-conversion" id="id14">Run the conversion</a></p></li>
<li><p><a class="reference internal" href="#inspect-the-dataset" id="id15">Inspect the dataset</a></p></li>
</ul>
</li>
</ul>
</nav>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a> function converts one or more VCF files to Zarr files stored in
sgkit’s Xarray data representation.</p>
<section id="highlights">
<h2><a class="toc-backref" href="#id1" role="doc-backlink">Highlights</a><a class="headerlink" href="#highlights" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Reads bgzip-compressed VCF and BCF files.</p></li>
<li><p>Large VCF files can be partitioned into regions using a Tabix (<code class="docutils literal notranslate"><span class="pre">.tbi</span></code>) or CSI (<code class="docutils literal notranslate"><span class="pre">.csi</span></code>)
index, and each region is processed in parallel using <a class="reference external" href="https://dask.org/">Dask</a>.</p></li>
<li><p>VCF parsing is performed by <a class="reference external" href="https://github.com/brentp/cyvcf2">cyvcf2</a>,
a Cython wrapper around <a class="reference external" href="https://github.com/samtools/htslib">htslib</a>,
the industry-standard VCF library.</p></li>
<li><p>Control over Zarr chunk sizes allows VCFs with a large number of samples
to be converted efficiently.</p></li>
<li><p>Input and output files can reside on local filesystems, Amazon S3, or
Google Cloud Storage.</p></li>
<li><p>Support for polyploid and mixed-ploidy genotypes.</p></li>
</ul>
</section>
<section id="installation">
<h2><a class="toc-backref" href="#id2" role="doc-backlink">Installation</a><a class="headerlink" href="#installation" title="Permalink to this heading">#</a></h2>
<p>VCF support is an “extra” feature within sgkit and requires additional
dependencies, notably <code class="docutils literal notranslate"><span class="pre">cyvcf2</span></code>.</p>
<p>To install sgkit with VCF support using pip (there is no conda package):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install &#39;sgkit[vcf]&#39;
</pre></div>
</div>
<p>There are <a class="reference external" href="https://github.com/brentp/cyvcf2#installation">installation instructions for cyvcf2</a>,
which may be helpful if you encounter errors during installation.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Reading VCFs is not supported on Windows, since <code class="docutils literal notranslate"><span class="pre">cyvcf2</span></code> and <code class="docutils literal notranslate"><span class="pre">htslib</span></code> do
not <a class="reference external" href="https://github.com/brentp/cyvcf2/issues/90">currently work on Windows</a>.
As a workaround, consider using scikit-allel’s <code class="docutils literal notranslate"><span class="pre">vcf_to_zarr</span></code> function
to write a VCF in Zarr format, followed by <a class="reference internal" href="generated/sgkit.read_scikit_allel_vcfzarr.html#sgkit.read_scikit_allel_vcfzarr" title="sgkit.read_scikit_allel_vcfzarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.read_scikit_allel_vcfzarr()</span></code></a> to
read the VCF as a <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2024.3.1.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a>.</p>
</div>
</section>
<section id="usage">
<h2><a class="toc-backref" href="#id3" role="doc-backlink">Usage</a><a class="headerlink" href="#usage" title="Permalink to this heading">#</a></h2>
<p>To convert a single VCF or BCF file to Zarr, just specify the input and output file names:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">sgkit.io.vcf</span> <span class="kn">import</span> <span class="n">vcf_to_zarr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vcf_to_zarr</span><span class="p">(</span><span class="s2">&quot;CEUTrio.20.21.gatk3.4.g.vcf.bgz&quot;</span><span class="p">,</span> <span class="s2">&quot;output.zarr&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s2">&quot;output.zarr&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:               (alleles: 4, ploidy: 2, samples: 1, variants: 19910)</span>
<span class="go">Dimensions without coordinates: alleles, ploidy, samples, variants</span>
<span class="go">Data variables:</span>
<span class="go">    call_genotype         (variants, samples, ploidy) int8 dask.array&lt;chunksize=(10000, 1, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_genotype_mask    (variants, samples, ploidy) bool dask.array&lt;chunksize=(10000, 1, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_genotype_phased  (variants, samples) bool dask.array&lt;chunksize=(10000, 1), meta=np.ndarray&gt;</span>
<span class="go">    sample_id             (samples) &lt;U7 dask.array&lt;chunksize=(1,), meta=np.ndarray&gt;</span>
<span class="go">    variant_allele        (variants, alleles) object dask.array&lt;chunksize=(10000, 4), meta=np.ndarray&gt;</span>
<span class="go">    variant_contig        (variants) int8 dask.array&lt;chunksize=(10000,), meta=np.ndarray&gt;</span>
<span class="go">    variant_id            (variants) object dask.array&lt;chunksize=(10000,), meta=np.ndarray&gt;</span>
<span class="go">    variant_id_mask       (variants) bool dask.array&lt;chunksize=(10000,), meta=np.ndarray&gt;</span>
<span class="go">    variant_position      (variants) int32 dask.array&lt;chunksize=(10000,), meta=np.ndarray&gt;</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:                    [&#39;20&#39;, &#39;21&#39;]</span>
<span class="go">    max_variant_allele_length:  48</span>
<span class="go">    max_variant_id_length:      1</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a> function can accept multiple files, and furthermore, each of these
files can be partitioned to enable parallel processing.</p>
</section>
<section id="multiple-files">
<h2><a class="toc-backref" href="#id4" role="doc-backlink">Multiple files</a><a class="headerlink" href="#multiple-files" title="Permalink to this heading">#</a></h2>
<p>If there are multiple files, then pass a list:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">sgkit.io.vcf</span> <span class="kn">import</span> <span class="n">vcf_to_zarr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vcf_to_zarr</span><span class="p">([</span><span class="s2">&quot;CEUTrio.20.gatk3.4.g.vcf.bgz&quot;</span><span class="p">,</span> <span class="s2">&quot;CEUTrio.21.gatk3.4.g.vcf.bgz&quot;</span><span class="p">],</span> <span class="s2">&quot;output.zarr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Processing multiple inputs is more work than a single file, since behind the scenes each input is
converted to a separate temporary Zarr file on disk, then these files are concatenated and rechunked
to form the final output Zarr file.</p>
<p>In the single file case, the input VCF is converted to the output Zarr file in a single sequential
pass with no need for intermediate temporary files. For small files this is fine, but for very large
files it’s a good idea to partition them so the conversion runs faster.</p>
</section>
<section id="partitioning">
<h2><a class="toc-backref" href="#id5" role="doc-backlink">Partitioning</a><a class="headerlink" href="#partitioning" title="Permalink to this heading">#</a></h2>
<p>Partitioning a large VCF file involves breaking it into a number of roughly equal-sized parts that can
be processed in parallel. The parts are specified using genomic regions that follow the regions format
used in <a class="reference external" href="http://samtools.github.io/bcftools/bcftools.html">bcftools</a>: <code class="docutils literal notranslate"><span class="pre">chr:beg-end</span></code>,
where positions are 1-based and inclusive.</p>
<p>All files to be partitioned must have either a Tabix (<code class="docutils literal notranslate"><span class="pre">.tbi</span></code>) or CSI (<code class="docutils literal notranslate"><span class="pre">.csi</span></code>) index. If both are present
for a particular file, then Tabix is used for finding partitions.</p>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.partition_into_regions.html#sgkit.io.vcf.partition_into_regions" title="sgkit.io.vcf.partition_into_regions"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.partition_into_regions()</span></code></a> function will create a list of region strings for a VCF
file, given a desired number of parts to split the file into:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">sgkit.io.vcf</span> <span class="kn">import</span> <span class="n">partition_into_regions</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">partition_into_regions</span><span class="p">(</span><span class="s2">&quot;CEUTrio.20.21.gatk3.4.g.vcf.bgz&quot;</span><span class="p">,</span> <span class="n">num_parts</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="go">[&#39;20:1-10108928&#39;, &#39;20:10108929-10207232&#39;, &#39;20:10207233-&#39;, &#39;21:1-10027008&#39;, &#39;21:10027009-10043392&#39;, &#39;21:10043393-10108928&#39;, &#39;21:10108929-10141696&#39;, &#39;21:10141697-10174464&#39;, &#39;21:10174465-10190848&#39;, &#39;21:10190849-10207232&#39;, &#39;21:10207233-&#39;]</span>
</pre></div>
</div>
<p>It’s important to note that the number of regions returned may not be exactly the number of parts
requested: it may be more or less. However, it is guaranteed that the regions will be contiguous and
will cover the whole VCF file.</p>
<p>The region strings are passed to <code class="docutils literal notranslate"><span class="pre">vcf_to_zarr</span></code> so it can process the parts in parallel:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">sgkit.io.vcf</span> <span class="kn">import</span> <span class="n">partition_into_regions</span><span class="p">,</span> <span class="n">vcf_to_zarr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">regions</span> <span class="o">=</span> <span class="n">partition_into_regions</span><span class="p">(</span><span class="s2">&quot;CEUTrio.20.21.gatk3.4.g.vcf.bgz&quot;</span><span class="p">,</span> <span class="n">num_parts</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vcf_to_zarr</span><span class="p">(</span><span class="s2">&quot;CEUTrio.20.21.gatk3.4.g.vcf.bgz&quot;</span><span class="p">,</span> <span class="s2">&quot;output.zarr&quot;</span><span class="p">,</span> <span class="n">regions</span><span class="o">=</span><span class="n">regions</span><span class="p">)</span>
</pre></div>
</div>
<p>It’s also possible to produce parts that have an approximate target size (in bytes). This is useful
if you are partitioning multiple files, and want all the parts to be roughly the same size.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">sgkit.io.vcf</span> <span class="kn">import</span> <span class="n">partition_into_regions</span><span class="p">,</span> <span class="n">vcf_to_zarr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">inputs</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CEUTrio.20.gatk3.4.g.vcf.bgz&quot;</span><span class="p">,</span> <span class="s2">&quot;CEUTrio.21.gatk3.4.g.vcf.bgz&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">regions</span> <span class="o">=</span> <span class="p">[</span><span class="n">partition_into_regions</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="n">target_part_size</span><span class="o">=</span><span class="mi">100_000</span><span class="p">)</span> <span class="k">for</span> <span class="nb">input</span> <span class="ow">in</span> <span class="n">inputs</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vcf_to_zarr</span><span class="p">(</span><span class="n">inputs</span><span class="p">,</span> <span class="s2">&quot;output.zarr&quot;</span><span class="p">,</span> <span class="n">regions</span><span class="o">=</span><span class="n">regions</span><span class="p">)</span>
</pre></div>
</div>
<p>The same result can be obtained more simply by specifying <code class="docutils literal notranslate"><span class="pre">target_part_size</span></code> in the call to
<code class="docutils literal notranslate"><span class="pre">vcf_to_zarr</span></code>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">sgkit.io.vcf</span> <span class="kn">import</span> <span class="n">vcf_to_zarr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">inputs</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CEUTrio.20.gatk3.4.g.vcf.bgz&quot;</span><span class="p">,</span> <span class="s2">&quot;CEUTrio.21.gatk3.4.g.vcf.bgz&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vcf_to_zarr</span><span class="p">(</span><span class="n">inputs</span><span class="p">,</span> <span class="s2">&quot;output.zarr&quot;</span><span class="p">,</span> <span class="n">target_part_size</span><span class="o">=</span><span class="mi">100_000</span><span class="p">)</span>
</pre></div>
</div>
<p>As a special case, <code class="docutils literal notranslate"><span class="pre">None</span></code> is used to represent a single partition.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">sgkit.io.vcf</span> <span class="kn">import</span> <span class="n">partition_into_regions</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">partition_into_regions</span><span class="p">(</span><span class="s2">&quot;CEUTrio.20.21.gatk3.4.g.vcf.bgz&quot;</span><span class="p">,</span> <span class="n">num_parts</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">None</span>
</pre></div>
</div>
</section>
<section id="chunk-sizes">
<h2><a class="toc-backref" href="#id6" role="doc-backlink">Chunk sizes</a><a class="headerlink" href="#chunk-sizes" title="Permalink to this heading">#</a></h2>
<p>One key advantage of using Zarr as a storage format is its ability to store
large files in chunks, making it straightforward to process the data in
parallel.</p>
<p>You can control the chunk <em>length</em> (in the variants dimension) and <em>width</em>
(in the samples dimension) by setting the <code class="docutils literal notranslate"><span class="pre">chunk_length</span></code> and <code class="docutils literal notranslate"><span class="pre">chunk_width</span></code>
parameters to <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a>.</p>
<p>Due to the way that VCF files are parsed, all of the sample data for a given
chunk of variants are loaded into memory at one time. In other words,
<code class="docutils literal notranslate"><span class="pre">chunk_length</span></code> is honored at read time, whereas <code class="docutils literal notranslate"><span class="pre">chunk_width</span></code> is honored
at write time. For files with very large numbers of samples, this can
exceed working memory. The solution is to also set <code class="docutils literal notranslate"><span class="pre">temp_chunk_length</span></code> to be a
smaller number (than <code class="docutils literal notranslate"><span class="pre">chunk_length</span></code>), so that fewer variants are loaded
into memory at one time, while still having the desired output chunk sizes
(<code class="docutils literal notranslate"><span class="pre">chunk_length</span></code> and <code class="docutils literal notranslate"><span class="pre">chunk_width</span></code>). Note that <code class="docutils literal notranslate"><span class="pre">temp_chunk_length</span></code> must
divide <code class="docutils literal notranslate"><span class="pre">chunk_length</span></code> evenly.</p>
</section>
<section id="cloud-storage">
<h2><a class="toc-backref" href="#id7" role="doc-backlink">Cloud storage</a><a class="headerlink" href="#cloud-storage" title="Permalink to this heading">#</a></h2>
<p>VCF files can be read from various file systems including cloud stores. However,
since different underlying libraries are used in different functions, there are
slight differences in configuration that are outlined here.</p>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.partition_into_regions.html#sgkit.io.vcf.partition_into_regions" title="sgkit.io.vcf.partition_into_regions"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.partition_into_regions()</span></code></a> function uses <a class="reference external" href="https://filesystem-spec.readthedocs.io/en/latest/">fsspec</a>
to read VCF metadata and their indexes. Therefore, to access files stored on Amazon S3 or Google Cloud Storage
install the <code class="docutils literal notranslate"><span class="pre">s3fs</span></code> or <code class="docutils literal notranslate"><span class="pre">gcsfs</span></code> Python packages, and use <code class="docutils literal notranslate"><span class="pre">s3://</span></code> or <code class="docutils literal notranslate"><span class="pre">gs://</span></code> URLs.</p>
<p>You can also pass <code class="docutils literal notranslate"><span class="pre">storage_options</span></code> to <a class="reference internal" href="generated/sgkit.io.vcf.partition_into_regions.html#sgkit.io.vcf.partition_into_regions" title="sgkit.io.vcf.partition_into_regions"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.partition_into_regions()</span></code></a> to configure the <code class="docutils literal notranslate"><span class="pre">fsspec</span></code> backend.
This provides a way to pass any credentials or other necessary arguments needed to <code class="docutils literal notranslate"><span class="pre">s3fs</span></code> or <code class="docutils literal notranslate"><span class="pre">gcsfs</span></code>.</p>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a> function does <em>not</em> use <code class="docutils literal notranslate"><span class="pre">fsspec</span></code>, since it
relies on <code class="docutils literal notranslate"><span class="pre">htslib</span></code> for file handling, and therefore has its own way of accessing
cloud storage. You can access files stored on Amazon S3 or Google Cloud Storage
using <code class="docutils literal notranslate"><span class="pre">s3://</span></code> or <code class="docutils literal notranslate"><span class="pre">gs://</span></code> URLs. Setting credentials or other options is
typically achieved using environment variables for the underlying cloud store.</p>
</section>
<section id="compression">
<h2><a class="toc-backref" href="#id8" role="doc-backlink">Compression</a><a class="headerlink" href="#compression" title="Permalink to this heading">#</a></h2>
<p>Zarr offers a lot of flexibility over controlling how data is compressed. Each variable can use
a different <a class="reference external" href="https://zarr.readthedocs.io/en/stable/tutorial.html#compressors">compression algorithm</a>,
and its own list of <a class="reference external" href="https://zarr.readthedocs.io/en/stable/tutorial.html#filters">filters</a>.</p>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a> function tries to choose good defaults for compression, using
information about the variable’s dtype, and also the nature of the data being stored.</p>
<p>For example, <code class="docutils literal notranslate"><span class="pre">variant_position</span></code> (from the VCF <code class="docutils literal notranslate"><span class="pre">POS</span></code> field) is a monotonically increasing integer
(within a contig) so it benefits from using a delta encoding to store the differences in its values,
since these are smaller integers that compress better. This encoding is specified using the NumCodecs
<a class="reference external" href="https://numcodecs.readthedocs.io/en/stable/delta.html">Delta</a> codec as a Zarr filter.</p>
<p>When converting from VCF you can specify the default compression algorithm to use for all variables
by specifying <code class="docutils literal notranslate"><span class="pre">compressor</span></code> in the call to <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a>. There are trade-offs
between compression speed and size, which this <a class="reference external" href="http://alimanfoo.github.io/2016/09/21/genotype-compression-benchmark.html">benchmark</a>
does a good job of exploring.</p>
<p>Sometimes you may want to override the compression for a particular variable. A good example of this
is for VCF FORMAT fields that are floats. Floats don’t compress well, and since there is a value for
every sample they can take up a lot of space. In many cases full float precision is not needed,
so it is a good idea to use a filter to transform the float to an int, that takes less space.</p>
<p>For example, the following code creates an encoding that can be passed to <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a>
to store the VCF <code class="docutils literal notranslate"><span class="pre">DS</span></code> FORMAT field to 2 decimal places. (<code class="docutils literal notranslate"><span class="pre">DS</span></code> is a dosage field that is between 0 and 2
so we know it will fit into an unsigned 8-bit int.):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">numcodecs</span> <span class="kn">import</span> <span class="n">FixedScaleOffset</span>

<span class="n">encoding</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;call_DS&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;filters&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">FixedScaleOffset</span><span class="p">(</span><span class="n">offset</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;f4&quot;</span><span class="p">,</span> <span class="n">astype</span><span class="o">=</span><span class="s2">&quot;u1&quot;</span><span class="p">)],</span>
    <span class="p">},</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Note that this encoding won’t work for floats that may be NaN. Consider using
<a class="reference external" href="https://numcodecs.readthedocs.io/en/stable/quantize.html">Quantize</a> (with <code class="docutils literal notranslate"><span class="pre">astype=np.float16</span></code>)
or <a class="reference external" href="https://numcodecs.readthedocs.io/en/stable/bitround.html">Bitround</a> in that case.</p>
</section>
<section id="low-level-operation">
<span id="vcf-low-level-operation"></span><h2><a class="toc-backref" href="#id9" role="doc-backlink">Low-level operation</a><a class="headerlink" href="#low-level-operation" title="Permalink to this heading">#</a></h2>
<p>Calling <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a> runs a two-step operation:</p>
<ol class="arabic simple">
<li><p>Write the output for each input region to a separate temporary Zarr store</p></li>
<li><p>Concatenate and rechunk the temporary stores into the final output Zarr store</p></li>
</ol>
<p>Each step is run as a Dask computation, which means you can use any Dask configuration
mechanisms to control aspects of the computation.</p>
<p>For example, you can set the Dask scheduler to run on a cluster. In this case you
would set the temporary Zarr store to be a cloud storage URL (by setting <code class="docutils literal notranslate"><span class="pre">tempdir</span></code>) so
that all workers can access the store (both for reading and writing).</p>
<p>For debugging, or for more control over the steps, consider using
<a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarrs.html#sgkit.io.vcf.vcf_to_zarrs" title="sgkit.io.vcf.vcf_to_zarrs"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarrs()</span></code></a> followed by <a class="reference internal" href="generated/sgkit.io.vcf.concat_zarrs.html#sgkit.io.vcf.concat_zarrs" title="sgkit.io.vcf.concat_zarrs"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.concat_zarrs()</span></code></a>.</p>
</section>
<section id="polyploid-and-mixed-ploidy-vcf">
<h2><a class="toc-backref" href="#id10" role="doc-backlink">Polyploid and mixed-ploidy VCF</a><a class="headerlink" href="#polyploid-and-mixed-ploidy-vcf" title="Permalink to this heading">#</a></h2>
<p>The <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a> function can be used to convert polyploid VCF
data to Zarr files stored in sgkit’s Xarray data representation by specifying the
ploidy of the dataset using the <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> parameter.</p>
<p>By default, sgkit expects VCF files to have a consistent ploidy level across all samples
and variants.
Manual specification of ploidy is necessary because, within the VCF standard,
ploidy is indicated by the length of each genotype call and hence it may not be
consistent throughout the entire VCF file.</p>
<p>If a genotype call of lower than specified ploidy is encountered it will be treated
as an incomplete genotype.
For example, if a VCF is being processed assuming a ploidy of four (i.e. tetraploid)
then the diploid genotype <code class="docutils literal notranslate"><span class="pre">0/1</span></code> will be treated as the incomplete tetraploid
genotype <code class="docutils literal notranslate"><span class="pre">0/1/./.</span></code>.</p>
<p>If a genotype call of higher than specified ploidy is encountered an exception is raised.
This exception can be avoided using the <code class="docutils literal notranslate"><span class="pre">truncate_calls</span></code> parameter in which case the
additional alleles will be skipped.</p>
<p>Conversion of mixed-ploidy VCF files is also supported by <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a>
by use of the <code class="docutils literal notranslate"><span class="pre">mixed_ploidy</span></code> parameter.
In this case <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> specifies the maximum allowed ploidy and lower ploidy
genotype calls within the VCF file will be preserved within the resulting dataset.</p>
<p>Note that many statistical genetics methods available for diploid data are not generalized
to polyploid and or mixed-ploidy data.
Therefore, some methods available in sgkit may only be applicable to diploid or fixed-ploidy
datasets.</p>
<p>Methods that are generalized to polyploid and mixed-ploidy data may make assumptions
such as polysomic inheritance and hence it is necessary to understand the type of polyploidy
present within any given dataset.</p>
</section>
<section id="example-converting-1000-genomes-vcf-to-zarr">
<h2><a class="toc-backref" href="#id11" role="doc-backlink">Example: converting 1000 genomes VCF to Zarr</a><a class="headerlink" href="#example-converting-1000-genomes-vcf-to-zarr" title="Permalink to this heading">#</a></h2>
<p>This section shows how to convert the <a class="reference external" href="https://www.internationalgenome.org/">1000 genomes</a> dataset into Zarr format for analysis in sgkit.</p>
<p>For reference, the conversion (not including downloading the data) took about an hour on a machine with 32 vCPUs and 128GB of memory (GCP e2-standard-32).</p>
<section id="install-sgkit">
<h3><a class="toc-backref" href="#id12" role="doc-backlink">Install sgkit</a><a class="headerlink" href="#install-sgkit" title="Permalink to this heading">#</a></h3>
<p>Install the main package using conda or pip, and the VCF extra package using pip, as described in <a class="reference internal" href="user_guide.html#installation"><span class="std std-ref">Installation</span></a>.</p>
</section>
<section id="download-the-data">
<h3><a class="toc-backref" href="#id13" role="doc-backlink">Download the data</a><a class="headerlink" href="#download-the-data" title="Permalink to this heading">#</a></h3>
<p>Run the following to download the 1000 genomes VCF files over FTP:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>mkdir -p data/1kg
for contig in {1..22}; do
  wget -P data/1kg ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr${contig}.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz
  wget -P data/1kg ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr${contig}.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz.tbi
done
</pre></div>
</div>
</section>
<section id="run-the-conversion">
<h3><a class="toc-backref" href="#id14" role="doc-backlink">Run the conversion</a><a class="headerlink" href="#run-the-conversion" title="Permalink to this heading">#</a></h3>
<p>Run the following Python code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sgkit.io.vcf</span> <span class="kn">import</span> <span class="n">vcf_to_zarr</span>
<span class="kn">from</span> <span class="nn">dask.distributed</span> <span class="kn">import</span> <span class="n">Client</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">client</span> <span class="o">=</span> <span class="n">Client</span><span class="p">(</span><span class="n">n_workers</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span> <span class="n">threads_per_worker</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">vcfs</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;data/1kg/ALL.chr</span><span class="si">{</span><span class="n">contig</span><span class="si">}</span><span class="s2">.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz&quot;</span> <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)]</span>
    <span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;1kg.zarr&quot;</span>
    <span class="n">vcf_to_zarr</span><span class="p">(</span><span class="n">vcfs</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">tempdir</span><span class="o">=</span><span class="s2">&quot;1kg-tmp&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>A few notes about the code:</p>
<p>1. Using a Dask distributed cluster, even on a single machine, performs better than the default scheduler (which uses threads), or
the multiprocessing scheduler. Creating a <code class="docutils literal notranslate"><span class="pre">Client</span></code> object will start a local cluster.</p>
<p>2. Making the number of workers less than the number of cores (16 rather than 32 in this case) will improve performance.
It’s  also important to set <code class="docutils literal notranslate"><span class="pre">threads_per_worker</span></code> to 1 to avoid overcommitting threads, as recommended in <a class="reference external" href="https://distributed.dask.org/en/latest/worker.html#thread-pool">the Dask documentation</a>.</p>
<p>3. It is useful to track the progress of the computation using <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-distributed.html#dashboard">the Dask dashboard</a>.
There are two steps in the conversion operation, described in <a class="reference internal" href="#vcf-low-level-operation"><span class="std std-ref">Low-level operation</span></a>, the first of which has coarse-grained, long-running tasks,
and the second which has much shorter-running tasks. There is a considerable delay (around 10 minutes) between the two steps,
so don’t worry if it doesn’t look like it’s progressing.</p>
<p>4. Only the core VCF fields and genotypes are converted. To import more VCF fields see the documentation
for the <code class="docutils literal notranslate"><span class="pre">fields</span></code> and <code class="docutils literal notranslate"><span class="pre">field_defs</span></code> parameters for <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html#sgkit.io.vcf.vcf_to_zarr" title="sgkit.io.vcf.vcf_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.vcf.vcf_to_zarr()</span></code></a>.</p>
</section>
<section id="inspect-the-dataset">
<h3><a class="toc-backref" href="#id15" role="doc-backlink">Inspect the dataset</a><a class="headerlink" href="#inspect-the-dataset" title="Permalink to this heading">#</a></h3>
<p>When the conversion is complete, have a look at the dataset as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s2">&quot;1kg.zarr&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:               (variants: 81271745, samples: 2504, ploidy: 2, alleles: 4)</span>
<span class="go">Dimensions without coordinates: variants, samples, ploidy, alleles</span>
<span class="go">Data variables:</span>
<span class="go">    call_genotype         (variants, samples, ploidy) int8 dask.array&lt;chunksize=(10000, 1000, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_genotype_mask    (variants, samples, ploidy) bool dask.array&lt;chunksize=(10000, 1000, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_genotype_phased  (variants, samples) bool dask.array&lt;chunksize=(10000, 1000), meta=np.ndarray&gt;</span>
<span class="go">    sample_id             (samples) object dask.array&lt;chunksize=(1000,), meta=np.ndarray&gt;</span>
<span class="go">    variant_allele        (variants, alleles) object dask.array&lt;chunksize=(10000, 4), meta=np.ndarray&gt;</span>
<span class="go">    variant_contig        (variants) int8 dask.array&lt;chunksize=(10000,), meta=np.ndarray&gt;</span>
<span class="go">    variant_id            (variants) object dask.array&lt;chunksize=(10000,), meta=np.ndarray&gt;</span>
<span class="go">    variant_id_mask       (variants) bool dask.array&lt;chunksize=(10000,), meta=np.ndarray&gt;</span>
<span class="go">    variant_position      (variants) int32 dask.array&lt;chunksize=(10000,), meta=np.ndarray&gt;</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:               [&#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;...</span>
<span class="go">    max_alt_alleles_seen:  12</span>
</pre></div>
</div>
</section>
</section>
</section>

<div class="section ablog__blog_comments">
  
  
</div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="user_guide.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">User Guide</p>
      </div>
    </a>
    <a class="right-next"
       href="examples/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Examples</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#highlights">Highlights</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#installation">Installation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#usage">Usage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-files">Multiple files</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#partitioning">Partitioning</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chunk-sizes">Chunk sizes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cloud-storage">Cloud storage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#compression">Compression</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#low-level-operation">Low-level operation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#polyploid-and-mixed-ploidy-vcf">Polyploid and mixed-ploidy VCF</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-converting-1000-genomes-vcf-to-zarr">Example: converting 1000 genomes VCF to Zarr</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#install-sgkit">Install sgkit</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#download-the-data">Download the data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#run-the-conversion">Run the conversion</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-the-dataset">Inspect the dataset</a></li>
</ul>
</li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">

  <div class="tocsection sourcelink">
    <a href="_sources/vcf.rst.txt">
      <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
  </div>
</div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2020, sgkit developers.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 6.2.1.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>